src/include/btree_cmp_inline.h

```diff
@@ -166,6 +300,143 @@ __wt_compare_bounds(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_ITEM *key, u
     return (0);
 }
 
+#ifdef HAVE_X86INTRIN_H
+/*
+ * __lex_compare_skip_gt_16 --
+ *     Lexicographic comparison routine for data greater than 16 bytes, skipping leading bytes.
+ *     Returns: < 0 if user_item is lexicographically < tree_item = 0 if user_item is
+ *     lexicographically = tree_item > 0 if user_item is lexicographically > tree_item We use the
+ *     names "user" and "tree" so it's clear in the btree code which the application is looking at
+ *     when we call its comparison function. Some platforms like ARM offer dedicated instructions
+ *     for reading 16 bytes at a time, allowing for faster comparisons. Some platforms like ARM
+ *     offer dedicated instructions for reading 16 bytes at a time, allowing for faster comparisons.
+ */
+static WT_INLINE int
+__lex_compare_skip_gt_16(
+  const uint8_t *ustartp, const uint8_t *tstartp, size_t len, int lencmp, size_t *matchp)
+{
+    __m128i res_eq, t, u;
+    uint64_t t64, tfirst, u64, ufirst;
+    int leading_zero_bytes;
+    size_t match;
+    const uint8_t *tendp, *treep, *uendp, *userp;
+    bool firsteq;
+
+    match = *matchp;
+    uendp = ustartp + len;
+    tendp = tstartp + len;
+
+    /*
+     * Compare 16 bytes at a time until we find a difference or run out of 16 byte chunks to
+     * compare.
+     */
+    for (userp = ustartp + match, treep = tstartp + match; uendp - userp > WT_VECTOR_SIZE;
+         userp += WT_VECTOR_SIZE, treep += WT_VECTOR_SIZE) {
+        u = _mm_loadu_si128((const __m128i *)userp);
+        t = _mm_loadu_si128((const __m128i *)treep);
+        res_eq = _mm_cmpeq_epi8(u, t);
+        if (_mm_movemask_epi8(res_eq) != 65535) {
+            match = (size_t)(userp - ustartp);
+            goto final128;
+        }
+    }
+
+    /*
+     * Rewind until there is exactly 16 bytes left. We know we started with at least 16, so we are
+     * still in bound.
+     */
+    u = _mm_loadu_si128((const __m128i *)(uendp - WT_VECTOR_SIZE));
+    t = _mm_loadu_si128((const __m128i *)(tendp - WT_VECTOR_SIZE));
+    match = (size_t)(uendp - ustartp) - WT_VECTOR_SIZE;
+
+final128:
+    ufirst = (uint64_t)_mm_extract_epi64(u, 0);
+    tfirst = (uint64_t)_mm_extract_epi64(t, 0);
+    firsteq = ufirst == tfirst;
+    u64 = firsteq ? (uint64_t)_mm_extract_epi64(u, 1) : ufirst;
+    t64 = firsteq ? (uint64_t)_mm_extract_epi64(t, 1) : tfirst;
+    match += firsteq ? sizeof(uint64_t) : 0;
+
+#ifndef WORDS_BIGENDIAN
+    u64 = __wt_bswap64(u64);
+    t64 = __wt_bswap64(t64);
+#endif
+
+    WT_LEADING_ZEROS(u64 ^ t64, leading_zero_bytes);
```

Comment:
This is silly. You have tried a bit too hard to match the arm code rather than using SSE-specific abilities. For one thing, x86 is always bigendian, so you don't need to check that. But really we should just do:

```c
res_eq = _mm_cmpeq_epi8(u, t);
uint32_t eq_bits = _mm_movemask_epi8(res_eq);
if (eq_bits = 0xFFFF) {
    // handle all bytes equal case
} else {
    WT_TRAILING_ZEROS(~eq_bits, leading_zero_bytes); // Adjust for 16-bit width if needed
}
```


It looks like glibc's [memcmp asm](https://github.com/bminor/glibc/blob/1ea051145612f199d8716ecdf78b084b00b5a727/sysdeps/x86_64/multiarch/memcmp-sse2.S#L181-L190) uses the equivalent of `leading_zero_bytes` to load the byte at that offset in each buffer and compare them, rather than trying to shuffle the data from the vectors to GPRs. I assume they have explored both strategies and found that to be faster, so I suggest doing the same.

Reply:
Thanks. Good idea. Fixed.

Reply:
x86 is little endian right? I still need to get 64bit of integer and flip them to compare. The good point is that I can return the equal case early.

Reply:
I tried your suggestion but it is too complicated to make it right. You have to consider eq_bits is also store in little endian and you need to flip it in x86.

Reply:
I tried your suggestion but there is still testing failing and I'm struggling to understand how. Appreciate if you can spot the problem easily.

Reply:
FYI `_mm_extract_epi64` was also [added in SSE4.1](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#ig_expand=0,2931&text=_mm_extract_epi64), so if we can use that, there is no problem with using `_mm_test_all_zeros`

(still looking into the correctness)